module/executiondatasync/pruner/pruner.go

```diff
@@ -0,0 +1,206 @@
+package pruner
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/executiondatasync/tracker"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+)
+
+const (
+	defaultHeightRangeTarget = uint64(400_000)
+	defaultThreshold         = uint64(100_000)
+)
+
+// Pruner is a component responsible for pruning data from
+// execution data storage. It is configured with the following
+// parameters:
+// * Height range target: The target number of most recent blocks
+//   to store data for. This controls the total amount of data
+//   stored on disk.
+// * Threshold: The number of block heights that we can exceed
+//   the height range target by before pruning is triggered. This
+//   controls the frequency of pruning.
+// The Pruner consumes a stream of fulfilled height notifications,
+// and triggers pruning once the difference between the fulfilled
+// height and the last pruned height reaches the height range
+// target + threshold.
+type Pruner struct {
+	storage       tracker.Storage
+	pruneCallback func(ctx context.Context) error
+
+	// channels used to send new fulfilled heights and config changes to the worker thread
+	// fulfilledHeights uses an unbounded channel to ensure NotifyFulfilledHeight is not blocking
+	fulfilledHeightsIn    chan<- interface{}
```

Comment:
I found the word `fulfilled` a bit confusing, I think `unpruned` might be easier to understand.

The reason I had confusion is because I thought there were three status: pruned, unpruned, fulfilled. 

I understand `fulfilled` means the height has been executed (fulfilled), but to the pruner's perspective, the only difference is whether a height is pruned or not. If it's not pruned, then it's unpruned. So there are only two status.